### PR TITLE
Fix for Issue #433

### PR DESCRIPTION
**Pull request by daminichopra**
## Proposed changes

Added NGSI-LD temporal property 'modifiedAt'  and 'observedAt'
Fix for [Issue](https://github.com/orchestracities/ngsi-timeseries-api/issues/433)

## Types of changes

What types of changes does your code introduce to the project: _Put
an `x` in the boxes that apply_

- [ x ] Bugfix (non-breaking change which fixes an issue)
- [ x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to
ask. We're here to help! This is simply a reminder of what we are going
to look for before merging your code._

- [ x ] I have read the [CONTRIBUTING][contrib] doc
- [ x ] I have signed the [CLA][cla]
- [ ] I have added tests that prove my fix is effective or that my
      feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in
      downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion
by explaining why you chose the solution you did and what alternatives
you considered, etc...

[cla]: https://raw.githubusercontent.com/orchestracities/ngsi-timeseries-api/master/individual-cla.pdf
    "Martel Open Source Software Individual Contributor License Agreement"
[contrib]: https://github.com/smartsdk/ngsi-timeseries-api/blob/master/CONTRIBUTING.md
    "Contributing to QuantumLeap"

### CODE REVIEW

File: src/reporter/delete.py [hunk 1]
```diff
@@ -34,11 +34,19 @@ def delete_entity(entity_id, type_=None, from_date=None, to_date=None):
 def delete_entities(entity_type, from_date=None, to_date=None,
                     drop_table=False):
     with translator_for(fiware_s()) as trans:
-        if drop_table:
```
Comment (c0c0n3) [inline src/reporter/delete.py:37]: it looks like this file doesn't belong in this PR? is it a relic of the other pr you worked on to drop tables? is git playing tricks on you? :-)

File: src/reporter/reporter.py [hunk 2]
```diff
@@ -90,7 +91,20 @@ def _validate_payload(payload):
 
     if 'id' not in payload:
         return 'Entity id is required in notifications'
-
+     # NGSI-LD temporal property
```
Comment (c0c0n3) [inline src/reporter/reporter.py:94]: the implementation should be in `reporter.timex` not here I think?

File: src/reporter/delete.py (hunk 1 above)
Comment (daminichopra) [reply src/reporter/delete.py:37]: Hi @c0c0n3 , As I have same repository and different branches for all previously raised PR that is why I guess its showing this [PR ](https://github.com/orchestracities/ngsi-timeseries-api/pull/455/files) changes too.

File: src/reporter/reporter.py (hunk 2 above)
Comment (daminichopra) [reply src/reporter/reporter.py:94]: Hi @c0c0n3 , I tried implementation in `reporter.timex`  but I faced some issues and after debugging code as per my understanding if `modifiedAt` or `observedAt` is in payload then it should be validated first so I implemented in `reporter.py` file. 
Please let me know if I am missing something. Thank you

Comment (chicco785) [reply src/reporter/reporter.py:94]: @daminichopra maybe the issue was not clear.

each time a payload arrives, a time_index is computed with a specific logic in `reporter.timex`, following the introduction of support for ngsi-ld, this logic needs to be updated.

so I am not sure that the PR so far complies with that.

**Review by chicco785 — CHANGES_REQUESTED** [review]

**Review by chicco785 — CHANGES_REQUESTED** [review]
i cleaned the branch, my comments about the current implementation holds

**Comment by chicco785** [comment]
@daminichopra i am not sure it's the right direction...

I will show you how to proceed over your branch, in the hope this is of help for future PRs:

1. Let's define the test cases and develop them. In this case @jason-fox was very kind and basically gave us the example to follow: https://github.com/orchestracities/ngsi-timeseries-api/issues/433#issuecomment-830071970. So let's develop the test case that based on the examples by json check that the time_index in the timeseries returned is the correct one.

-> I will make a first pr adding these test cases 

2. let's modify `time_index_priority_list` in `timex.py` to set as timeindex the selected attributes.

-> I will make a second pr with the changes required

3. if all goes well after this PR test cases define in 1 will pass.

File: docs/manuals/user/using.md [hunk 3]
```diff
@@ -297,6 +297,20 @@ in any of the attribute metadata section in the notification.
 As specified in the
 [FIWARE data models documentation](https://fiware-datamodels.readthedocs.io/en/latest/guidelines/index.html#dynamic-attributes).
 
+1. **observedAt** NGSI-LD metadata. The most recent `observedAt` attribute value
+found in any of the attributes in the notification.
+As specified in the
+[NGSI-LD](https://github.com/smart-data-models/data-models/blob/master/ngsi-ld_howto.md#steps-to-migrate-to-json-ld).
```
Comment (chicco785) [inline docs/manuals/user/using.md:303]: @jason-fox any better reference than this?

**Comment by chicco785** [comment]
> @daminichopra i am not sure it's the right direction...
> 
> I will show you how to proceed over your branch, in the hope this is of help for future PRs:
> 
> 1. Let's define the test cases and develop them. In this case @jason-fox was very kind and basically gave us the example to follow: [Include NGSI-LD observed/modified props in time index calcs #433 (comment)](https://github.com/orchestracities/ngsi-timeseries-api/issues/433#issuecomment-830071970). So let's develop the test case that based on the examples by json check that the time_index in the timeseries returned is the correct one.
> 
> -> I will make a first pr adding these test cases
> 
> 1. let's modify `time_index_priority_list` in `timex.py` to set as timeindex the selected attributes.
> 
> -> I will make a second pr with the changes required
> 
> 1. if all goes well after this PR test cases define in 1 will pass.

@daminichopra let me know if you understand what I did and if we can agree to this approach for your future PRs

File: docs/manuals/user/using.md (hunk 3 above)
Comment (jason-fox) [reply docs/manuals/user/using.md:303]: Direct from the [NGSI-LD Specification](https://www.etsi.org/deliver/etsi_gs/CIM/001_099/009/01.04.01_60/gs_cim009v010401p.pdf)

>
> In case of conflicting information for an Attribute, where a `datasetId` is duplicated, but there are differences in the other attribute data, the one with the most recent `observedAt` DateTime, if present, and otherwise the one with the most recent `modifiedAt` DateTime shall be provided. 

Strictly this quote is about multi-attributes, but the precedence order is always the same

Comment (chicco785) [reply docs/manuals/user/using.md:303]: i was hoping for a web spec, but that's life :)

Comment (chicco785) [reply docs/manuals/user/using.md:303]: also, it seems from `https://ngsi-ld-tutorials.readthedocs.io/en/latest/ngsi-ld-operations.html` that `modifiedAt` can be attached to entity itself and not just the attribute (despite from the examples it seems that this value would be the same as the last modification of an attribute).

**Comment by daminichopra** [comment]
> > @daminichopra i am not sure it's the right direction...
> > I will show you how to proceed over your branch, in the hope this is of help for future PRs:
> > 
> > 1. Let's define the test cases and develop them. In this case @jason-fox was very kind and basically gave us the example to follow: [Include NGSI-LD observed/modified props in time index calcs #433 (comment)](https://github.com/orchestracities/ngsi-timeseries-api/issues/433#issuecomment-830071970). So let's develop the test case that based on the examples by json check that the time_index in the timeseries returned is the correct one.
> > 
> > -> I will make a first pr adding these test cases
> > 
> > 1. let's modify `time_index_priority_list` in `timex.py` to set as timeindex the selected attributes.
> > 
> > -> I will make a second pr with the changes required
> > 
> > 1. if all goes well after this PR test cases define in 1 will pass.
> 
> @daminichopra let me know if you understand what I did and if we can agree to this approach for your future PRs

@chicco785 , yes i got it. I will keep this approach in mind. Thanks for this.

**Comment by chicco785** [comment]
@jason-fox i think this is ready for landing, do you mind having a look and check if you requirements have been implemented correctly? (in which case you can approve the pr). if you think some other ngsi-ld experts are better to be involved, let me know!

File: docs/manuals/user/using.md (hunk 3 above)
Comment (jason-fox) [reply docs/manuals/user/using.md:303]: This is `options=sysAttrs` - the `modifiedAt` is generated by the Context Broker for each attribute and for the entity as a whole. I think there could be a divergence if an entity has multiple attributes and only some are updated since the unmodified attributes would have an older `modifiedAt` timestamp.

Obviously this won't be an issue for a subscribing component since the notification will only be raised when an attribute changes.

File: src/reporter/tests/test_notify.py [hunk 4]
```diff
@@ -695,3 +695,151 @@ def test_json_ld(service, notification):
...
+        QL_URL)
+    url_new = '{}'.format(get_url)
+    r = requests.post(url, data=json.dumps(notification),
+                      headers=notify_header(service))
+    assert r.status_code == 200
+    # Give time for notification to be processed
+    time.sleep(SLEEP_TIME)
+    res_get = requests.get(url_new, headers=query_header(service))
+    assert res_get.status_code == 200
+    assert_equal_time_index_arrays([res_get.json()['index'][0]], ["2021-01-28T12:33:10.000Z"])
```
Comment (jason-fox) [inline src/reporter/tests/test_notify.py:781]: This doesn't seem to match the `observedAt` in line 766

**Review by jason-fox — APPROVED** [review]
I just have one query but the overall logic looks right.

File: docs/manuals/user/using.md (hunk 3 above)
Comment (chicco785) [reply docs/manuals/user/using.md:303]: @jason-fox thanks! is this documented in any official doc of ngsi-ld or a related implementation? I think it would be good to refer to it in the internal ql docs

File: src/reporter/tests/test_notify.py (hunk 4 above)
Comment (chicco785) [reply src/reporter/tests/test_notify.py:781]: indeed...
